### ISSUE TITLE
accuracy difference in local inference and triton inference

### ISSUE THREAD

**Issue by inzamamulDU** [triton-inference-server/server#4512]
I am new in triton. so I am investigating how triton works.

For this, I made inference both triton and local for mnist dataset. but in local inference I got accuracy is 95.20% and triton inference server got 99.70%. 

I followed same pre-process for image pre processing . and the backend I used is TensorFlow .

can anyone told me that? is triton can increases accuracy ?

**Comment by dyastremsky** [comment]
Did you use the same (frozen) model on the same data? Triton can increase latency, throughput, functionality, and other attributes, but it should not impact accuracy. That depends on the model and data.

**Comment by inzamamulDU** [comment]
@dyastremsky  yes, I use the same model on same data. I also checked the test data is same or not. and also the preprocessed data. 

they are same.  below this is my local inference code in client side
`import cv2, random, requests
import numpy as np
from glob import glob
from tqdm import tqdm

TEST_MNIST_PATHS = glob("/Users/inzamamul/Downloads/archive/MNIST - JPG - testing/*/*.jpg")
MODEL_ENDPOINT = "http://0.0.0.0:5000/mnist_infer"
NUM_SAMPLES = 1000
INPUT_SHAPE = (28, 28)

r = random.Random()
r.seed('Hi')

# Choose random images from the test set
path_choices = r.choices(TEST_MNIST_PATHS, k=NUM_SAMPLES)
#print(path_choices)


def preprocess_image(image):
    image = cv2.resize(image, INPUT_SHAPE) # This step is not strictly required for MNIST
    image = image / 255 # Normalize Image
    return image.astype(np.float32)

accuracy = list()

# Loop though each image and run inference
for test_mnist_path in tqdm(path_choices):
    label = int(test_mnist_path.split("/")[-2]) # Get the label
    image = cv2.imread(test_mnist_path, -1) # Read image without converting to BGR
    image = preprocess_image(image) # Preprocess the image

    # Encode the image as JPG and Send to the Model Server
    _, img_encoded = cv2.imencode('.jpg', image)
    response = requests.post(MODEL_ENDPOINT, data=img_encoded.tobytes())
    pred = int(response.json()) # Decode the response to get the predictions


    accuracy.append(pred == label)
print(f"Testing Accuracy: {np.mean(accuracy)*100:.2f}%")
`

and this is my server code. 

`from flask import Flask, redirect, jsonify, request
import cv2, os, sys, imutils
import numpy as np
import tensorflow as tf

MODEL_PATH = 'mnist_model'

app = Flask(__name__)
app.config['DEBUG'] = False

# Load the Model
model = tf.keras.models.load_model(MODEL_PATH)

@app.route('/mnist_infer', methods=['POST'])
def hand_classifier():
    # Receive the encoded frame and convert it back to a Numpy Array
    encoded_image = np.frombuffer(request.data, np.uint8)
    image = cv2.imdecode(encoded_image, -1) # Decode image without converting to BGR

    image = np.expand_dims([image], axis=-1) # Add dimensions to create appropriate tensor shapes

    # Run inference on the frame
    hand = model.predict(image)

    return jsonify(str(np.argmax(hand))) # Because only string can be converted to JSON

if __name__ == '__main__':
    app.run(host='0.0.0.0', port='5000', threaded=True, use_reloader=False)
`

and here this is my triton inference code .

`import cv2, random
import numpy as np
from glob import glob
from tqdm import tqdm
import tritonclient.grpc as grpcclient

# GLOBAL VARIABLES
TEST_MNIST_PATHS = glob("/Users/inzamamul/Downloads/archive/MNIST - JPG - testing/*/*.jpg")
NUM_SAMPLES = 1000
INPUT_SHAPE = (28, 28)

# Triton Variables
TRITON_IP = "localhost"
TRITON_PORT = 8001
MODEL_NAME = "mnist_model"
INPUTS = []
OUTPUTS = []
INPUT_LAYER_NAME = "input_1"
OUTPUT_LAYER_NAME = "output_1"

r = random.Random()
r.seed('Hi')

# Choose random images from the test set
path_choices = r.choices(TEST_MNIST_PATHS, k=NUM_SAMPLES)
#print(path_choices)


def preprocess_image(image):
    image = cv2.resize(image, INPUT_SHAPE) # This step is not strictly required for MNIST
    image = image / 255 # Normalize Image
    image = np.expand_dims([image], axis=-1) # Increase the dimensions to match that of the model
    return image.astype(np.float32)

def postprocess_output(preds):
    return np.argmax(np.squeeze(preds))

accuracy = list()

# Triton Initializations
INPUTS.append(grpcclient.InferInput(INPUT_LAYER_NAME, [1, INPUT_SHAPE[0], INPUT_SHAPE[1], 1], "FP32"))
OUTPUTS.append(grpcclient.InferRequestedOutput(OUTPUT_LAYER_NAME))
TRITON_CLIENT = grpcclient.InferenceServerClient(url=f"{TRITON_IP}:{TRITON_PORT}")

# Loop though each image and run inference
for test_mnist_path in tqdm(path_choices):
    label = int(test_mnist_path.split("/")[-2]) # Get the label
    image = cv2.imread(test_mnist_path, -1) # Read image without converting to BGR
    image = preprocess_image(image) # Preprocess the image
    #print(image)

    # Run the Inference using the gRPC triton client
    INPUTS[0].set_data_from_numpy(image) # Set the Inputs
    result = TRITON_CLIENT.infer(model_name=MODEL_NAME, inputs=INPUTS, outputs=OUTPUTS, headers={}) # Run Inference
    output = np.squeeze(result.as_numpy(OUTPUT_LAYER_NAME)) # Process the Outputs
    pred = postprocess_output(output) # Postprocess (Argmax)


    # Record Accuracy
    accuracy.append(pred == label)
#print(accuracy)
print(f"Testing Accuracy: {np.mean(accuracy)*100:.2f}%")
`

can you help me out .

**Comment by dyastremsky** [comment]
It looks like your code has a few spots where there are implicit type conversions (e.g. between int and fp32), which you may want to look at first. MNIST data is typically stored as integers in the file.

Try printing out every tensor/matrix along the inference for a couple of images to see if the intermediary tensors/matrices are getting changed or differ. That'll give more information about where the difference could be occurring.

**Comment by inzamamulDU** [comment]
this is the last some image data from local inference 
<img width="525" alt="Screen Shot 2022-06-14 at 11 48 50 PM" src="https://user-images.githubusercontent.com/38895016/173657694-473f65c4-bcba-4638-9f27-17f2ee8e8a15.png">

and this is the exact same image data from triton inference 
<img width="192" alt="Screen Shot 2022-06-14 at 11 51 09 PM" src="https://user-images.githubusercontent.com/38895016/173657782-be4c986f-a78f-40fb-b3cd-689545b1d735.png">


I did this before and saw they are equal. I cannot find the problem .

**Comment by dyastremsky** [comment]
There can also be variance across TensorFlow versions. Does your Python script use the same Python TensorFlow image as the TensorFlow backend you are using (we have two backends, not sure which you are using)? Loading with Keras could theoretically impose some slight differences in internal ops or model execution compared to our TensorFlow backend as well.

**Comment by dyastremsky** [comment]
One more thing: have you tried testing your script in the TensorFlow container that corresponds to the version of Triton you are testing with: nvcr.io/nvidia/tensorflow:YY.MM-py3? That would give you the TensorFlow Python wheel used in the corresponding Triton container (same YY.MM).

**Comment by inzamamulDU** [comment]
> docker run --rm -it -p 8000-8002:8000-8002 --name triton_server -v $PWD:/models nvcr.io/nvidia/tritonserver:21.09-py3 tritonserver --model-repository=/models --strict-model-config=false --backend-config=tensorflow,version=1

I ran the command for using TensorFlow backend with TensorFlow version 1.x and 2.x both as --backend-config=tensorflow,version=<int> CLI argument 

but both cases I got same accuracy 99.70% 

I don't know this command line is okay or not .

> There can also be variance across TensorFlow versions. Does your Python script use the same Python TensorFlow image as the TensorFlow backend you are using (we have two backends, not sure which you are using)?

my TensorFlow version is **1.16.0** .  I tested it also by giving this argument --backend-config=tensorflow,version=<int> in that command.

**Comment by dyastremsky** [comment]
Thanks for your response. I'm not seeing 1.16.0 in the TensorFlow official releases: https://www.tensorflow.org/versions.

Can you try the latest Triton container, 22.05? Also, could you try running your script in the TensorFlow container that corresponds to the Triton version you are using (nvcr.io/nvidia/tensorflow:YY.MM-py3, mentioned in the last comment). This will help identify whether it has to do with the TensorFlow version (i.e. outside of Triton) or inside the server.

**Comment by inzamamulDU** [comment]
I changed my TensorFlow version to 2.9.1 from 1.6.0 


and run docker using this command for TensorFlow backend running 

`docker run --rm -it -p 8000-8002:8000-8002 --name triton_server -v $PWD:/models nvcr.io/nvidia/tritonserver:21.09-py3 tritonserver --model-repository=/models --strict-model-config=false --backend-config=tensorflow,version=1`

but still I got the same accuracy

**Comment by inzamamulDU** [comment]
can you tell me the exact command to run TensorFlow container ?

**Comment by dyastremsky** [comment]
Run your local inference script in the container.

**Comment by inzamamulDU** [comment]
I cannot run the local inference script in the container, can you please tell me how to run ?

**Comment by inzamamulDU** [comment]
this is the last 2 images result probability from my local inference 
```
[[9.6399826e-06 2.7133070e-16 1.4669935e-07 5.9463789e-09 5.9044527e-08
  2.0892419e-05 3.6993213e-05 3.4245599e-13 9.9993110e-01 1.1801465e-06]] 
[[7.2736685e-11 1.9384597e-08 8.5478598e-01 8.7181054e-02 4.5359680e-09
  6.9063915e-08 3.7449532e-09 8.3604959e-08 5.8032818e-02 1.7986974e-08]]
```

and this is the last 2 images result probability from triton inference
```
[3.1131749e-06 1.2712182e-16 4.4254961e-07 2.7573155e-06 8.3282997e-10
 2.4722605e-05 8.6542841e-06 1.0097854e-12 9.9995923e-01 1.0560907e-06]
[2.6409364e-11 5.2485518e-08 9.9954730e-01 2.6126110e-04 4.3995363e-09
 5.9927969e-09 3.4451140e-11 7.4019837e-08 1.9123880e-04 6.2072083e-08]
```

**Comment by inzamamulDU** [comment]
I solved the problem, the problem is actually passing the tensor from client to server in local inference 

for this I just passed the image without encoding. 

thanks for the effort to you. I closed this issue.

**Comment by monalvision** [comment]
@inzamamulDU , can you please tell me what encoding changes you have done to fix this? I am facing the same issue

**Comment by inzamamulDU** [comment]
@monalvision what kind of issue are you exactly facing? can you describe me a little bit ?

**Comment by inzamamulDU** [comment]
@monalvision 
is the onnx model file is same as pytorch or tensorflow. There is a opset_version which should be matched in the onnx model generation and in the inference script. 

From my case, in the pytorch case, i made a script to convert the pytorch model to onnx in this flow 

PyTorch model to TorchScript, you can use the torch.jit.trace or torch.jit.script functions and then torchscript to onnx by classyfing the opset_version properly. 

For tensorflow, there is a library for converting "tf2onnx".

**Comment by monalvision** [comment]
I have exported the model using yolov5 export.py and serving that onnx in triton server for inference.

triton-onnx inference output
tensor([[0.0000e+00, 9.2163e+00, 1.4484e+03, 9.2008e+02, 6.5887e-01, 1.0000e+00]])

yolov5-official-gpu
tensor([[4.60000e+01, 2.60000e+02, 1.41900e+03, 9.23000e+02, 5.45649e-01, 0.00000e+00]], device='cuda:0')

yolov5-official-cpu
tensor([[4.60000e+01, 2.60000e+02, 1.41900e+03, 9.22000e+02, 5.45604e-01, 0.00000e+00]])

**Comment by dyastremsky** [comment]
Please don't tag me. We don't provide support on closed issues. If you would like support for your new issue, please create a new issue using the bug template and provide full reproduction steps.

In this case, I'd also recommend running your inference outside of Triton using the same version of your model framework to validate that the issue is not in the framework itself.
